### PR TITLE
Add `map_length` sql function

### PR DESCRIPTION
**Pull request by guswynn**
Broken out of https://github.com/MaterializeInc/materialize/pull/11012, helps testing multiple things. A dual of `list_length`

### Motivation

  * This PR adds a feature that has not yet been specified.
see above

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add a `map_length` function to give the size of a `map` type, similar to `list_length`.
  - Change the return type of the `list_length` function from `bigint` to `integer` for parity with `array_length`.

### CODE REVIEW

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -1351,6 +1351,10 @@ fn map_get_values<'a>(a: Datum<'a>, b: Datum<'a>, temp_storage: &'a RowArena) ->
     })
 }
 
+fn map_length(a: Datum) -> Datum {
+    Datum::Int64(a.unwrap_map().iter().count() as i64)
```
Comment (benesch) [inline src/expr/src/scalar/func.rs:1355]: This is going to be incorrect at the boundary. (I realize it's copied from `list_length`—that one is busted too!) Need to use `try_into`, and produce an IntOutOfRange` error if it fails. 

Btw, I think for consistency with `array_length` we should make these return `i32`. Technically you can create a list or map with more than 2^32 elements, but you're going to have a _very_ bad time if you do. So I think it's okay to change the return type of the length functions to `i32`, and then file a follow up issue to _enforce_ that you can't create an array/map/list with more than 2^32 issues.

File: doc/user/content/sql/types/map.md [hunk 2]
```diff
@@ -236,3 +236,16 @@ SELECT '{a=>1, b=>2}'::map[text=>int] ?| ARRAY['c', 'd', '1'] as search_for_any_
 ---------------------
  f
 ```
+
+#### Count entries in map (`map_length)
+
+Returns the number of entries in the map
```
Comment (benesch) [inline doc/user/content/sql/types/map.md:242]:
```suggestion
Returns the number of entries in the map.
```

File: src/expr/src/scalar/func.rs (hunk 1 above)
Comment (guswynn) [reply src/expr/src/scalar/func.rs]: good point! updated the pr to clean everything up!

File: src/expr/src/scalar/func.rs [hunk 3]
```diff
@@ -2726,9 +2733,9 @@ impl BinaryFunc {
             ))
             .nullable(true),
 
-            ListLengthMax { .. } | ArrayLength | ArrayLower | ArrayUpper => {
-                ScalarType::Int64.nullable(true)
-            }
+            ArrayLength | ArrayLower | ArrayUpper => ScalarType::Int64.nullable(true),
```
Comment (benesch) [inline src/expr/src/scalar/func.rs:2736]: Oh noo, these should all be Int32 too. This PR is ballooning in scope, so feel free to file a follow-up issue for this one instead.

**Review by benesch — APPROVED** [review]

File: src/expr/src/scalar/func.rs (hunk 3 above)
Comment (guswynn) [reply src/expr/src/scalar/func.rs:2736]: Ill merge as-is, and cleanup the array ones in a later pr!

Comment (benesch) [reply src/expr/src/scalar/func.rs:2736]: Sounds good. I filed #11527 to track!

Comment (guswynn) [reply src/expr/src/scalar/func.rs:2736]: https://github.com/MaterializeInc/materialize/pull/11532

**Pull request merged**